### ISSUE TITLE
`tf.lite.OpsSet.SELECT_TF_OPS` makes Interpreter invoke function crash

### ISSUE THREAD

**Issue by tgeffroyModuleus** [tensorflow/tensorflow#56946]
### System information

-   **Have I written custom code (as opposed to using a stock example script
    provided in TensorFlow)**: yes
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 20.04
-   **TensorFlow installed from (source or binary)**: Binary (Anaconda)
-   **TensorFlow version (use command below)**: 2.4.1
-   **Python version**: 3.9.12
-   **cudatoolkit version**: 10.1.243
-   **cuDNN version**: 7.6.5
-   **GPU model and memory**: NVIDIA GeForce GTX 1650 Mobile / Max-Q
-  **nvidia-smi**:
```
> Fri Jul 29 21:31:26 2022       
> +-----------------------------------------------------------------------------+
> | NVIDIA-SMI 515.48.07    Driver Version: 515.48.07    CUDA Version: 11.7     |
> |-------------------------------+----------------------+----------------------+
> | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
> | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
> |                               |                      |               MIG M. |
> |===============================+======================+======================|
> |   0  NVIDIA GeForce ...  Off  | 00000000:01:00.0 Off |                  N/A |
> | N/A   55C    P8     2W /  N/A |      4MiB /  4096MiB |      0%      Default |
> |                               |                      |                  N/A |
> +-------------------------------+----------------------+----------------------+
>                                                                                
> +-----------------------------------------------------------------------------+
> | Processes:                                                                  |
> |  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
> |        ID   ID                                                   Usage      |
> |=============================================================================|
> |    0   N/A  N/A      1185      G   /usr/lib/xorg/Xorg                  4MiB |
> +-----------------------------------------------------------------------------+

```
-   **Anaconda-env**: [anaconda-env.zip](https://github.com/tensorflow/tensorflow/files/9222799/anaconda-env.zip)

### Describe the problem
Hello, I am starting working with Tensorflow and I am facing a problem that seems similar to this issue: https://github.com/tensorflow/tensorflow/issues/25231

#### Context
I have a trained model I cannot really change and I have to deploy it on Android.  
My first thought was to use Tensorflow Lite to convert it and use it on Android. So I have converted it and I have tried to load it on my computer first (before trying to deploy it on Android). While the `invoke` function of the interpreter is called, it crashes (or the execution is aborted) with no error.  
It seems it cannot handle 5D tensor as [Oktai15](https://github.com/Oktai15) says but I cannot find any documentation that explains this constraint. Moreover, even if it cannot handle the case, it should return an error message and it should be explained in the documentation.  

In order to be sure the problem is the fact that the interpreter doesn't handle 5D tensors, I have created a simple script that trains a model on the equation `y = 2*x`.

#### Source code working with 4D tensor
Here is the code that use a 4D-tensor to predict the result for the equation `y = 2*x`

```
import pathlib
import numpy as np
import matplotlib.pyplot as plt

import tensorflow as tf

print('\u2022 Using TensorFlow Version:', tf.__version__)

# Create a simple Keras model.
x = [-5, -1, 0, 1, 4]
y = [-10, -2, 0, 2, 8]

model = tf.keras.models.Sequential([
        tf.keras.layers.Dense(units = 1, input_shape=[1,1,1,1])
])
model.compile(optimizer='sgd',
              loss='mean_squared_error')

model.fit(x, y, epochs=200)

export_dir = 'saved_model/2'
tf.saved_model.save(model, export_dir)

# Convert the model.
converter = tf.lite.TFLiteConverter.from_saved_model(export_dir)
# converter.target_spec.supported_ops = [
#   tf.lite.OpsSet.SELECT_TF_OPS # enable TensorFlow ops.
# ]
tflite_model = converter.convert()

tflite_model_file = pathlib.Path('model.tflite')
tflite_model_file.write_bytes(tflite_model)

# Load TFLite model and allocate tensors.
interpreter = tf.lite.Interpreter(model_content=tflite_model)
interpreter.allocate_tensors()

# Get input and output tensors.
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

# Test the TensorFlow Lite model on random input data.
input_shape = input_details[0]['shape']
inputs, outputs, outputsTFL = [], [], []
for _ in range(100):
    input_data = np.array(np.random.random_sample(input_shape)*100, dtype=np.float32)
    interpreter.set_tensor(input_details[0]['index'], input_data)
    
    interpreter.invoke()
    tflite_results = interpreter.get_tensor(output_details[0]['index'])
    
    # Test the TensorFlow model on random input data.
    tf_results = model(tf.constant(input_data))
    output_data = np.array(tf_results)
    
    inputs.append(input_data[0][0][0][0][0])
    outputs.append(output_data[0][0][0][0][0])
    outputsTFL.append(np.array(tflite_results)[0][0][0][0][0])

plt.plot(inputs, outputs, 'r')
plt.show()
plt.plot(inputs, outputsTFL, 'r')
plt.show()
```

#### Source code not working with 5D tensor
Here is the code that use a 5D-tensor to predict the result for the equation `y = 2*x`.   

```
import pathlib
import numpy as np
import matplotlib.pyplot as plt

import tensorflow as tf

print('\u2022 Using TensorFlow Version:', tf.__version__)

# Create a simple Keras model.
x = [-5, -1, 0, 1, 4]
y = [-10, -2, 0, 2, 8]

model = tf.keras.models.Sequential([
        tf.keras.layers.Dense(units = 1, input_shape=[1,1,1,1,1])
])
model.compile(optimizer='sgd',
              loss='mean_squared_error')

model.fit(x, y, epochs=200)

export_dir = 'saved_model/2'
tf.saved_model.save(model, export_dir)

# Convert the model.
converter = tf.lite.TFLiteConverter.from_saved_model(export_dir)
converter.target_spec.supported_ops = [
  tf.lite.OpsSet.SELECT_TF_OPS # enable TensorFlow ops.
]
tflite_model = converter.convert()

tflite_model_file = pathlib.Path('model.tflite')
tflite_model_file.write_bytes(tflite_model)

# Load TFLite model and allocate tensors.
interpreter = tf.lite.Interpreter(model_content=tflite_model)
interpreter.allocate_tensors()

# Get input and output tensors.
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

# Test the TensorFlow Lite model on random input data.
input_shape = input_details[0]['shape']
inputs, outputs, outputsTFL = [], [], []
for _ in range(100):
    input_data = np.array(np.random.random_sample(input_shape)*100, dtype=np.float32)
    interpreter.set_tensor(input_details[0]['index'], input_data)
    
    interpreter.invoke()
    tflite_results = interpreter.get_tensor(output_details[0]['index'])
    
    # Test the TensorFlow model on random input data.
    tf_results = model(tf.constant(input_data))
    output_data = np.array(tf_results)
    
    inputs.append(input_data[0][0][0][0][0][0])
    outputs.append(output_data[0][0][0][0][0][0])
    outputsTFL.append(np.array(tflite_results)[0][0][0][0][0][0])

plt.plot(inputs, outputs, 'r')
plt.show()
plt.plot(inputs, outputsTFL, 'r')
plt.show()
```

**WARNING**: The converter needs the quantization that enables Tensorflow ops: `tf.lite.OpsSet.SELECT_TF_OPS`. Otherwise, this error is raised:   
```<unknown>:0: error: loc(callsite(callsite("sequential/dense/BiasAdd@__inference__wrapped_model_1628" at "StatefulPartitionedCall@__inference_signature_wrapper_1742") at "StatefulPartitionedCall")): 'tf.BiasAdd' op is neither a custom op nor a flex op
<unknown>:0: note: loc("StatefulPartitionedCall"): called from
<unknown>:0: error: failed while converting: 'main': Ops that can be supported by the flex runtime (enabled via setting the -emit-select-tf-ops flag):
	tf.BiasAdd {data_format = "NHWC", device = ""}
````

#### Source code not working with 1D tensor
After trying a new configuration, it seems the problem comes from `tf.lite.OpsSet.SELECT_TF_OPS`  
When it is passed to the  converter even the configuration with 1D tensor is not working:
```
model = tf.keras.models.Sequential([
        tf.keras.layers.Dense(units = 1, input_shape=[1])
])
```

**Thanks**

_Note: I will not be able to execute code until august 17th._

**Comment by mohantym** [comment]
Hi @tgeffroyModuleus !

Input shape has to be = [1] instead of [1,1,1,1,1] (i.e one output for one input) in model side.
Attached resolved [gist](https://colab.sandbox.google.com/gist/mohantym/36338e8ef7361fa0cb50d7d44d567918/git_56946.ipynb) for reference. 

Thank you!

**Comment by tgeffroyModuleus** [comment]
Hi,  
The purpose of the code was to test 5D tensor in a very simple code sample. I know the input shape [1] is working without `tf.lite.OpsSet.SELECT_TF_OPS`, it is not the point here.
If you have a better sample to propose which takes 5D tensor I will be pleased to test it.
However it doesn't explain either why it is working with input shape [1,1,1,1] and not with [1,1,1,1,1]. 
You also could test input shape [1] but with `tf.lite.OpsSet.SELECT_TF_OPS`
Thanks,

**Comment by mohantym** [comment]
@tgeffroyModuleus !
Thanks for the clarification. Yes! 5D tensors are not supported yet in Lite side.
Proper syntax to use select ops syntax is 
```
converter.target_spec.supported_ops = [
  tf.lite.OpsSet.TFLITE_BUILTINS, # enable TensorFlow Lite ops.
  tf.lite.OpsSet.SELECT_TF_OPS # enable TensorFlow ops.
]
```
Attached update [gist](https://colab.research.google.com/gist/mohantym/36338e8ef7361fa0cb50d7d44d567918/git_56946.ipynb?pli=1#scrollTo=1oUhj0PJrbYF) for reference.
Thank you!

**Comment by tgeffroyModuleus** [comment]
Ok, thanks!  
Is there a documentation that explains this constraint please?  
Whis is tensorflow lite crashing without any error message?

**Comment by mohantym** [comment]
@tgeffroyModuleus !
Actually I can get results after using np.expand_dims with input dataset.  But the results were not as expected. Attached [gist](https://colab.research.google.com/gist/manas1988/e83eb4be6fe862c3ebcfb82bec9c4642/git_56946.ipynb#scrollTo=S7-G8rXjWt33) for reference.  5D might be  partially supported for linear regression. Could not find the exact documentation on 5D and 6D tensors .

But got some pointers here. [1](https://github.com/tensorflow/tensorflow/issues/56808#issuecomment-1188969571) , [2](https://github.com/tensorflow/tensorflow/issues/42923#issuecomment-687046286) .

@sachinprasadhs ! Could you look at this issue.

Thank you!

**Comment by sachinprasadhs** [comment]
@tgeffroyModuleus , Thanks for reporting the issue. As of now it is not documented about the support on 5D tensors, we will look into that.
Meanwhile, you could try with 4D tensors, alternatively you could try flattening the array and then reshaping it back to 5D tensors after conversion?

**Comment by tgeffroyModuleus** [comment]
Thank you, I am looking forward to learn more about your tensor support and constraints. 
I am not able to look at "my" neural network code right now. I will see what is possible to do when I get back (August 17th), but as I have explained in the issue, I am not owner of the neural network code, it can be difficult for me to make changes accepted on it.

Is there an other way than Tensorflow Lite to deploy a trained model on Android?

**Comment by tgeffroyModuleus** [comment]
In order to find a solution to deploy a trained model using 5D tensors on Android I have opened this question on stackoverflow:
https://stackoverflow.com/questions/73474682/how-to-deploy-a-tensorflow-model-on-android-without-tf-lite-and-without-using-a

**Comment by sachinprasadhs** [comment]
TFLite now supports 5D in selectV2, could you please try in the latest nightly version or the Tensorflow version 2.10 and confirm. More details can be found here https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md#major-features-and-improvements-1

**Comment by tgeffroyModuleus** [comment]
The very basic sample seems to work while using np.expand_dims.

However, I have others operator in my model, and I cannot test the solution right now. I have tried to install my packages with this new version of tensorflow and packages tensorflow-base, keras-base, and keras-gpu cannot be installed with this latest version.

Without them I cannot launch my model that uses other operators:
- Conv3D
- MaxPooling3D
- Dropout
- Dense
- GlobalAveragePooling3D

**Comment by sachinprasadhs** [comment]
Sure, After you test with the new setup, please let us know the progress. Thanks!

**Comment by pjpratik** [comment]
Hi @tgeffroyModuleus 

Did you get a chance to check with your setup?

 I have tested with TF 2.12 with 5D tensor along with `SELECT_TF_OPS` and was able to run without any error.

Please find the gist [here](https://colab.research.google.com/gist/pjpratik/c4247a7f00481d6bdeb06515dc2b1636/56946.ipynb) and let us know if it helps.

Thanks.
